custom_components/kia_uvo/services.py

```diff
@@ -0,0 +1,73 @@
+import logging
+from typing import Any, cast
+
+
+from homeassistant.const import ATTR_DEVICE_ID
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import ServiceCall, callback, HomeAssistant
+from .coordinator import HyundaiKiaConnectDataUpdateCoordinator
+from homeassistant.helpers import device_registry
+
+from .const import DOMAIN
+
+SERVICE_UPDATE = "update"
+SERVICE_FORCE_UPDATE = "force_update"
+
+SUPPORTED_SERVICES = (SERVICE_UPDATE, SERVICE_FORCE_UPDATE)
+
+_LOGGER = logging.getLogger(__name__)
+
+@callback
+def async_setup_services(hass: HomeAssistant) -> None:
+    """Set up services for Hyundai Kia Connect"""
+
+    async def async_handle_force_update(call):
+        coordinator = _get_coordinator_from_device(hass, call)
+        await coordinator.async_force_update_all()
+    
+
+    async def async_handle_update(call):
+        coordinator = hass.data[DOMAIN][config_entry.unique_id]
+        await coordinator.async_update_all()
+        
+
+    services = {
+        SERVICE_FORCE_UPDATE: async_handle_force_update,
+        SERVICE_UPDATE: async_handle_update,
+    }
+    
+    for service in SUPPORTED_SERVICES:
+        hass.services.async_register(
+            DOMAIN,
+            service,
+            services[service]
+        )
+
+@callback
+def async_unload_services(hass) -> None:
+    for service in SUPPORTED_SERVICES:
+        hass.services.async_remove(DOMAIN, service)
+
+
+
+def _get_coordinator_from_device(hass: HomeAssistant, call: ServiceCall) -> HyundaiKiaConnectDataUpdateCoordinator:
+    dev_reg = device_registry.async_get(hass)
+    device_entry = dev_reg.async_get(call.data[ATTR_DEVICE_ID][0])
+    _LOGGER.debug(
+        f"Device: {device_entry}"
+    )
+    config_entry_ids = device_entry.config_entries
+    config_entry_id = next(
```

Comment:
> > This is none functional. I could use a hand figuring out the _get_coordinator_from_device method in services.py.
> 
> is this still valid?

No it works now.

Reply:
I found it in the link you posted.   I don't fully understand it but it works.  I believe it is looking to find the entry id because it returns a list not a single item.